core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java

```diff
@@ -398,25 +428,57 @@ public synchronized void write(byte[] bs, int off, int len)
 	}
 
 	// Implement timeout with a scheduled task
-	ScheduledFuture<Object> sf = null;
+	ScheduledFuture<String> sf = null;
 	try {
-	    try {
-		if (timeout > 0)
-		    sf = ses.schedule(timeoutTask,
-					timeout, TimeUnit.MILLISECONDS);
-	    } catch (RejectedExecutionException ex) {
-		// ignore it; Executor was shut down by another thread,
-		// the following write should fail with IOException
-	    }
-	    os.write(bs, off, len);
+        try {
+            if (timeout > 0)
+                sf = ses.schedule(timeoutTask,
+                    timeout, TimeUnit.MILLISECONDS);
+        } catch (RejectedExecutionException ex) {
+            if (!socket.isClosed()) {
+                throw new IOException("Write aborted due to timeout not enforced", ex);
+            }
+        }
+
+        try {
+            os.write(bs, off, len);
+        } catch (IOException e) {
+            e = sf == null ? e : new IOException(handleTimeoutTaskResult(sf), e);
+            throw e;
+        }
 	} finally {
 	    if (sf != null)
-		sf.cancel(true);
-	}
+		    sf.cancel(true);
+	    }
     }
 
     @Override
     public void close() throws IOException {
 	os.close();
     }
+
+    private String handleTimeoutTaskResult(ScheduledFuture<String> sf)  {
+        boolean wasInterrupted = Thread.interrupted();
+        String exceptionMessage = null;
+        try {
+            if (!sf.cancel(true)) {
+                return sf.get(0, TimeUnit.NANOSECONDS);
+            }
```

Comment:
@jmehrens thank you for your review! I'm on vacation this week and will apply the suggested changes next week as soon as I'm back

Reply:
Hello @jmehrens , I'm back, sorry again for the delay. Your points make sense, just one question about the last one, do you suggest having something like that (not pass original IOException when sf.cancel is true):

```       
        try {
            os.write(bs, off, len);
        } catch (IOException e) {
            if (sf != null && sf.cancel(true)) {
                throw new IOException("The write wasn't completed due to write timeout");
            }
            e = sf == null ? e : new IOException(handleTimeoutTaskResult(sf), e);
            throw e;
        }
```
?

Reply:
Hello!  No worries on the delay.  I think what you have is correct.  I think we can get rid of some null checking if we write it like:
```
try {
            os.write(bs, off, len);
        } catch (IOException e) {
            if (sf != null && !sf.cancel(true)) {
                throw new IOException(handleTimeoutTaskResult(sf), e)
            }
            throw e;
        }
```

If we are able to

Reply:
but in this case, the caller will see a message like "Socket closed" without being aware that the write timeout occurred. 
Maybe it would be useful to provide somehow this info, what do you think?

Reply:
My understanding is: If `sf` is null there is no detailed result to fetch so throw original exception.  If we are able to cancel `sf` it will never run therefore therefore we are under the timeout threshold and there is no detailed result to fetch.  In that case we just throw the original exception.

Reply:
ah, right, my bad - got confused, thanks

Reply:
fixed